src/bbd/geocoder/geocoder.py

```diff
@@ -0,0 +1,582 @@
+    >>> geocode("1315 10th St, Sacramento, CA 95814")
+    Location(California State Capitol, 1315, 10th Street, Land Park, Sacramento,
+     Sacramento County, California, 95814, United States of America, (38.57656885,
+     -121.4934010890531, 0.0))
+
+    >>> email = "not_an_email"
+    >>> geocode = get_geocoder(email)
+    AssertionError: Must enter a valid email
+    """
+    #Must enter a valid email
+    assert is_valid_email(email), "Must enter a valid email"
+    geolocator=Nominatim(user_agent=email)
+    geocoder=RateLimiter(geolocator.geocode, min_delay_seconds=1)
+    return geocoder
+
+
+def get_reverse_geocoder(email):
+    """Get reverse geocode function for reverse geocoding through Nominatim 
+    with supplied email address as the user_agent, as per Nominatim's 
+    usage policy.
+
+    This reverse geocoder will take at least 1 second per Address queried
+    in accordance with Nominatim's terms of service.
+    Note: this process cannot be multi-threaded or run in parallel.
+
+    Arguments
+    ---------
+    email : str
+        A string email Address supplied to Nominatim as user_agent
+
+    Examples
+    --------
+
+    >>> email = "valid.email@Address.com"
+    >>> reverse_coder = get_reverse_geocoder(email)
+    >>> reverse_coder("37.873621099999994, -122.25768131042068")
+    Location(Evans Hall, Bechtel Drive, Northside, Berkeley, Alameda County,
+     California, 94720, United States of America, (37.873621099999994,
+     -122.25768131042068, 0.0))
+
+    >>> email = "not_an_email"
+    >>> reverse_coder = get_reverse_geocoder(email)
+    AssertionError: Must enter a valid email
+    """
+    #Must enter a valid email
+    assert is_valid_email(email), "Must enter a valid email"
+    geolocator = Nominatim(user_agent=email)
+    reverse_geocoder = RateLimiter(geolocator.reverse, min_delay_seconds=1)
+    return reverse_geocoder
+
+
+def street_encode(street) -> str:
+    """Default street_encoder for GeocodeLocation.
+
+    Gets street address while removing secondary unit designators.
+     See: https://pe.usps.com/text/pub28/28apc_003.htm
+    Not all secondary unit designators are removed as they may be contained
+     in street names e.g. key, hanger, pier
+
+    Arguments
+    ---------
+    street : str
+        a string street component of address
+
+    >>> street_encode("123 Test St")
+    "123 Test St"
+    >>> street_encode("456 Test Way apt 375")
+    "456 Test Way"
+    >>> street_encode("420 Test Ave Unit 5")
+    "420 Test Ave"
+    """
+    pattern = r"^.*\S(?=(?:\s+)(?:[Aa]p(?:ar)?t(?:ment)? |[Uu]nit |#|[Ll]ot |"\
+              r"[Rr](?:oo)?m |[Bb](?:ui)?ld(?:in)?g |[Uu]pp?e?r|[Ll]o?we?r|"\
+              r"[Ss](?:ui)?te |[Ff](?:l )?(?:rnt)?|[Tt]r(?:ai)?le?r |[Dd]ept ))"
+    try:
+        result = re.search(pattern, street) or street
+    except TypeError:
+        result = ""
+    if type(result) == str:
+        return result #No match
+    else:
+        return result[0] #Match
```

Comment:
I'm a little confused about the two sequential if/thens you're referring to? In this case the `try/except` block catches when there is a string that breaks regex (like bytecode if that made its way in ... actually maybe I should take out that `try/except` block). The `if/else` here returns the original string if there was no match (`type(result) == str`) and otherwise it takes apart the re.Match object.

Originally I wrote this block as `if type(result) is re.Match/else` but python 3.6 doesn't have re.Match objects instead its sre._sreMatch or something to that effect. Then I switched it to check for string instead for <=3.6 support